vro-streamlit/src/vro_streamlit/auth/auth_frontend.py

```diff
@@ -0,0 +1,59 @@
+import logging
+
+import streamlit as st
+
+import vro_streamlit.auth.auth_service as auth
+from vro_streamlit.auth.user import User
+
+AUTH_LOG_IN_BUTTON = 'auth_log_in_button'
+AUTH_LOG_OUT_BUTTON = 'auth_log_out_button'
+
+
+def log_out() -> None:
+    if st.button('Log Out', key=AUTH_LOG_OUT_BUTTON):
+        try:
+            auth.log_out(st.session_state.user)
+            st.success('Logged out successfully.')
+        except Exception as e:
+            logging.error(f'Failed to revoke token, but logged out anyways: {e}')
+
+        st.session_state.user = None
+        st.rerun()
+
+
+def log_in() -> None:
+    if st.button('Log In with GitHub', key=AUTH_LOG_IN_BUTTON, help='Log in to access more features.'):
+        # Step 1: Initiate the Device Flow
+        device_flow_data = auth.initiate_device_flow()
+
+        # Display the user code and verification URI
+        st.write(f'1. Go to [GitHub]({device_flow_data.verification_uri}) and enter the following code:')
+        st.code(device_flow_data.user_code)
+        st.write(f'2. This code expires in {device_flow_data.expires_in // 60} minutes.')
+
+        # Step 2: Poll for Access Token
+        with st.spinner('Waiting for authorization...'):
+            try:
+                access_token = auth.poll_for_token(device_flow_data.device_code, device_flow_data.interval)
+                try:
+                    # Step 3: Fetch User Info
+                    user_info = auth.fetch_user_info(access_token)
+                    st.success(f'Login successful! Logged in as: {user_info.login}')
+                    st.session_state.user = User(access_token, user_info.login, user_info.avatar_url)
+                    st.rerun()
+                except Exception as e:
+                    st.error(f'Could not gather user information: {e}')
+            except Exception as e:
+                st.error(f'Login failed: {e}')
+                st.session_state.user = None
```

Comment:
Marginal/Debatable: Instead of comments, we could break `log_in` down into smaller functions for each log_in step (the below is not tested and is pseudocode):
```python
def initiate_login_flow() -> DeviceFlowResponse:
    """Initiate the login flow and return device flow data."""
    return auth_service.initiate_device_flow()


def display_authorization_instructions(device_flow_data: DeviceFlowResponse) -> None:
    """Display instructions for the user to authorize the application."""
    st.write("### Authorization Steps")
    st.write("1. Visit [GitHub's verification page]({}).".format(device_flow_data.verification_uri))
    st.write("2. Enter the code:")
    st.code(device_flow_data.user_code, language='markdown')
    st.write("3. Complete the authorization within **{} minutes**.".format(device_flow_data.expires_in // 60))


def complete_login_flow(device_flow_data: DeviceFlowResponse) -> None:
    """Complete the login flow by polling for token and fetching user info."""
    try:
        access_token = auth_service.poll_for_token(device_flow_data.device_code, device_flow_data.interval)
        user_info = auth_service.fetch_user_info(access_token)
        st.success(f'Login successful! Logged in as: {user_info.login}')
        st.session_state.user = User(access_token, user_info.login, user_info.avatar_url)
        st.experimental_rerun()
    except Exception as e:
        logging.exception("An error occurred during login.")
        st.error("Login failed. {}".format(str(e)))
        st.session_state.user = None


def log_in() -> None:
    if st.button('Log In with GitHub', key='auth_log_in_button', help='Log in to access more features.'):
        try:
            device_flow_data = initiate_login_flow()
            display_authorization_instructions(device_flow_data)
            with st.spinner('Waiting for authorization...'):
                complete_login_flow(device_flow_data)
        except Exception as e:
            logging.exception("An error occurred during login initiation.")
            st.error("Login failed. {}".format(str(e)))
            st.session_state.user = None
```